design/sorted-tree.md

```diff
@@ -0,0 +1,166 @@
+
+Take this simple sorted tree example:
+
+```
+              root
++-------------------------------+
+|  +-----------+  +-----------+ |
+|  | 1, block1 |  | 4, block2 | |
+|  +-----------+  +-----------+ |
++-------------------------------+
+           
+     block1            block2
++-------------+    +-------------+
+| +---+ +---+ |    | +---+ +---+ |
+| | 1 | | 2 | |    | | 4 | | 5 | |
+| +---+ +---+ |    | +---+ +---+ |
++-------------+    +-------------+
+```
+
+Here we have a small and simple balanced tree using a chunking function that holds only 2 nodes per block.
+
+If we want to insert `3` into the structure we're confronted with several problems. If we append `block1` or
+prepend to `block2` you'll end up with a different tree shape, and a different root hash, than if you created
+the structure anew using the chunking function.
+
+**Rule: all sorted tree's must be consisently chunked in order to produce consistent tree structures which then
+produce consistent hashes.** This means that hash consistent tree will, in effect, be self-balancing upon mutation.
+
+If we stick with our "only 2 nodes per block" chunker then minor mutations to our list will produce mutations
+to the entire tree on the right side of that mutation. This is obviously not ideal.
+
+What we need is a new chunking technique that produces nodes of a desirable length but also consistently splits
+on particular **entries**. If we can find a way to consistently split on particular entries then we can avoid
+large mutations to the rest of the tree.
+
+## Chunker
+
+The chunker is a simple state machine. You feed the chunker unique identifiers for each entry. Each identifier must
+use derived from the entry, typically through a hashing function or from a digest already present in the entry.
+
+The chunker should break as consistently as possible while also protecting against various attacks (see below).
+
+## First Tree: Sorted CID Set (Tail Chunker)
+
+```ipldschema
+type Entry link
+type Leaf [ Entry ]
+type Branch struct {
+  start Link
+  link Link
+} representation tuple
+
+type Node union {
+  | "leaf" Leaf
+  | "branch" Branch
+} representation keyed
+```
+
+At this point we're ready to build our first tree. This data structure is a `Set` of CIDs sorted by binary comparison.
+The nice thing about working with this use case is each CID is both the *key* and the *value*.
+
+CID's end in a multihash, and the multihash typically ends in a hash digest. For this use case, we'll say that CID's that are allowed
+in this `Set` are limited to those that use sufficiently secure hashing functions.
+
+Since hash digests are, effectively, a form of randomization, we can simply convert the tail of each digest to an fixed size integer and designate
+some part of that address space to cause splits in our chunker.
+
+For simplicity, let's convert the last byte to Uint8. Now, let's make the chunker close every chunk when it sees a `0`. This will give
```

Comment:
Is "closing a chunk" an expression I don't know? If not, I think "closing a chunk" needs a definition.